src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -1795,7 +1795,12 @@ private void EmitInstanceCallExpression(BoundCall call, UseKind useKind)
 
             EmitCallCleanup(call.Syntax, useKind, method);
 
-            FreeOptTemp(tempOpt);
+            // We cannot free (and later reuse) receiver temp if the method itself is ref
+            // and hence could return reference to the temp.
+            if (method.RefKind == RefKind.None)
```

Comment:
>if (method.RefKind == RefKind.None) [](http://example.com/codeflow?start=12&length=35)

This doesn't feel like a right fix. I think scenarios in #67435 should produce ref safety errors instead. Sorry if my comment for one of the scenarios led to a confusion.

We probably should think about this situation as follows:
When we are calling a method on a struct value, we are introducing a new local to act as the receiver, and we are introducing a scope for that local that is the invocation. References to the temp are not allowed to escape its scope. So, if the method can return a reference to the receiver (with whatever means, return, ref/out parameter, etc.), that should be a ref safety error. I am not sure how to put it in spec-like wording, but perhaps @jaredpar and @RikkiGibson could help with that.

Like in the following example we report an error:
[https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUEEsC2UECeAwgPYB2yMCcAxjCQsgDQwjVkA+AAgEwAERALAAoAN4i+kvlwCMANj4BlPgFkZACgCUfALwA+PmQgB3JVr6i+AMV18Z9vgF8A3CIlTZCrGRiqe6pDA+b18ADyYlPgJNd0lxYSlE6QB2PlDXBKlHN0zJT2CfVQBmLViLMsSuVJV/QNUNTQA6ACUIMC0IgBM2qDgAGxhNDMTs4TL8upDVUtzy2cTlGAgcAAdbNS1hpLzUuqXVlrbNvgB6E74IBAQGAUUADnk7kAEoMjISXyq+PpIaKD6+AByfYrQF8YAEPiBS4QMg0CDgiB/ODIBFYXxYZCGD58KBQtrfX7/MqjRLjeTSAAsqig3hmiXi2zyMkasgAnOoVCVNEMSSJRjkuEU+JRqHQlCJGR5hVMrBkygBtWSNAAiWCgAHN3pQsDRkI1SN0AIJkf4EZCYxoAVQoNBIKwgnVaYAAuuNhZNCs76VIpdsvnU5XzhI4gA===](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUEEsC2UECeAwgPYB2yMCcAxjCQsgDQwjVkA+AAgEwAERALAAoAN4i+kvlwCMANj4BlPgFkZACgCUfALwA+PmQgB3JVr6i+AMV18Z9vgF8A3CIlTZCrGRiqe6pDA+b18ADyYlPgJNd0lxYSlE6QB2PlDXBKlHN0zJT2CfVQBmLViLMsSuVJV/QNUNTQA6ACUIMC0IgBM2qDgAGxhNDMTs4TL8upDVUtzy2cTlGAgcAAdbNS1hpLzUuqXVlrbNvgB6E74IBAQGAUUADnk7kAEoMjISXyq+PpIaKD6+AByfYrQF8YAEPiBS4QMg0CDgiB/ODIBFYXxYZCGD58KBQtrfX7/MqjRLjeTSAAsqig3hmiXi2zyMkasgAnOoVCVNEMSSJRjkuEU+JRqHQlCJGR5hVMrBkygBtWSNAAiWCgAHN3pQsDRkI1SN0AIJkf4EZCYxoAVQoNBIKwgnVaYAAuuNhZNCs76VIpdsvnU5XzhI4gA===)

```
class C
{
    static S M1() => new S() { F = 111 };

    static int M2(ref int x, S y)
    {
        return x;
    }

    static int M3()
    {
        return M2(ref M1().Ref(), default);
    }

    static ref int M()
    {
        S temp = M1();
        return ref temp.Ref(); // error CS8168: Cannot return local 'temp' by reference because it is not a ref local
    }
    
    static void Main()
    {
        System.Console.WriteLine(M3());
    }
}


public struct S
{
    public int F;

    [System.Diagnostics.CodeAnalysis.UnscopedRef]
    public ref int Ref()
    {
        return ref F;
    }
} 
```

Reply:
Also consider this example ([SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMACAYQFgAoAb1LyrywEYA2PAZTwFlaAKASjwF4A+PDACmAd2bc8ZPADE+eWorwBfANylK1Oo1gI2ODrDy68ADzjM8ATy6aqFEtSc0A7GfWPqyjZ6rbjMHqsAMzcdlLhTlhurAZG7NwAdABKwgBm3BYAJukQqAA2CFweTt4k4f4mIcBhvhF1UTFxMHipaUwIEAhQAMYcCVxc2bkFRSVedhUMNAAsbNAwtU4Ozlq0iXQAnP2hg+OrdBu029XcxeFlTlOMyOl4txBZAPYw+VYBem0dXb2GLSwAZx4AnudwBiRkHjKpCwwTwAIQaB6eiYpBWWjhJkhPicAG1DgARKAQADmMCeCN64IITxyAEEYBA3gCoOCAKowAE9J4AB2EWTaAF1wg9nq93rDQWlQY8Xm8Pq10ktqOiDm5btLsXUysogA===)):

```cs
class C
{
    static S M1() => new S() { F = 111 };

    static int M2(in int x, S y)
    {
        return x;
    }

    static int M3()
    {
        return M2(in M1().Ref(), default);
    }

    static int M3b()
    {
        return M2(in RefStatic(M1()), default);
    }
   
    static void Main()
    {
        System.Console.WriteLine(M3());
        System.Console.WriteLine(M3b());
    }
    
    static ref readonly int RefStatic(in S s) => ref s.F;
}

public struct S
{
    public int F;

    [System.Diagnostics.CodeAnalysis.UnscopedRef]
    readonly public ref readonly int Ref()
    {
        return ref F;
    }
}
```

It's the readonly sample from the issue with added method `M3b` which I think is equivalent to the existing method `M3` - I only "refactored" the instance method `S.Ref` into a static method `RefStatic`. But you are suggesting `M3` should be disallowed, whereas `M3b` should stay allowed?

Reply:
> But you are suggesting M3 should be disallowed, whereas M3b should stay allowed?

I am not suggesting that. I think that should be an error due to similar line of thinking and a small refactoring demonstrates that.

```
    static int M3b1()
    {
        return M2(in getRef(), default);
        
        static ref readonly int getRef()
        {
            // error CS8156: An expression cannot be used in this context because it may not be passed or returned by reference
            // error CS8347: Cannot use a result of 'C.RefStatic(in S)' in this context because it may expose variables referenced by parameter 's' outside of their declaration scope
            return ref RefStatic(M1());
        }
    }
```

The temp introduced for an `in` parameter is not guaranteed to "survive" past the invocation, at least I do not think language rules provide any guarantees around that (feel free to prove me wrong though).

I am not sure at the moment why the temp is not getting reused for the specific scenario, this could be due to an unrelated issue.

Reply:
> I am not sure at the moment why the temp is not getting reused for the specific scenario, this could be due to an unrelated issue.

I managed to break this scenario as well:
https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMACAYQFgAoAb1LyrywEYA2PAZTwFlaAKASjwF4A+PDACmAd2bc8ZPADE+eWorwBfANylK1Ooyb5WODrDywEeAB5dNVCiWp2aAdiFjmBi+tvVlGz1W3GYU1YAZmBuKykIu108AGd8XjYDIwAlYQAzJgQIBCgAYw52bi4uD3tqDhFxJmKAOgAVAHss5FgAc25VPAB6bsJGsDBhQLwEAAsoWLwAG1hhPEbUUwRGuOF5iDxgYTGIADcoRuQ8PN2YNuE4LaXRianZkTixxemAEyFG00gAaw2YAE88K8oOl0sJkMM8sIolonPFajIcGUvBEIv4sAAWNjQGDhXyRfF2Oi1OgATkKoWKyKo3nxaIYeAh6UZwggr0aMGmgJMeDSmWyuQKRhYsR4AhZzNiCI8tNIWGCcQQaDypiYpBsRIVPJkMp8TMVytVOHVaIV+ohbI5XICpkR1LwptchhgNvMlnxGvKsgSEvM9tpyiAA
```
class C
{
    static S M1() => new S() { F = 111 };

    static S2 M2(in int x)
    {
        return new S2(x);
    }

    static int M3b()
    {
        S2 s2 = M2(in RefStatic(M1()));
        (new S()).ToString(); // Comment this line out to see a behavior change, but this line should not make any difference
        return s2.F2;
    }
    
    static void Main()
    {
        System.Console.WriteLine(M3b());
    }
    
    static ref readonly int RefStatic(in S s) => ref s.F;
}

public struct S
{
    public int F;
}

ref struct S2
{
    public ref readonly int F2;
    
    public S2(in int x)
    {
        F2 = ref x;
    }
}
```